### ISSUE TITLE
Image.LoadAsync(stream) uses synchronous IO if the stream is seekable 

### ISSUE THREAD

**Issue by kroymann** [SixLabors/ImageSharp#1997]
### ImageSharp version

1.0.4 (verified issue is still present in 2.0.0)

### Other ImageSharp packages and versions

ImageSharp.Web 1.0.5 (+ fixes from PR 222)

### Environment (Operating system, version and so on)

Azure App Service (Windows-based)

### .NET Framework version

6.0

### Description

`Image.LoadAsync(stream)` ends up synchronously reading from the source stream in both `Image.InternalDetectFormatAsync()` and then again in `ImageDecoderUtilities.DecodeAsync()`.  When running in a webservice under high-load scenarios, this can quickly lead to thread pool exhaustion as all threads are blocked waiting on network IO.

```cs
private static Task<IImageFormat> InternalDetectFormatAsync(Stream stream, Configuration config)
{
    return Task.FromResult(InternalDetectFormat(stream, config));
}
```

```cs
public static Task<Image<TPixel>> DecodeAsync<TPixel>(...)
{
    using var bufferedReadStream = new BufferedReadStream(configuration, stream);
    Image<TPixel> image = decoder.Decode<TPixel>(bufferedReadStream, cancellationToken);
    return Task.FromResult(image);
}
```

### Steps to Reproduce

Try to load an image asynchronously and observe that threads end up blocked on synchronous reads

**Comment by kroymann** [comment]
I initially started to implement a PR that would fix this by making both of those methods _actually_ async, but that has long tendrils that extend deep into the image decoding code.  And even if you ignore the amount of duplicate sync/async code paths such a PR would create, I'm now questioning whether that's actually even the right way to fix this.

As a bit of background, this issue only occurs if the source stream passed in to `Image.LoadAsync()` is seekable, because in that scenario the image decoding reads directly from the source stream.  If the source stream is _not_ seekable, then the code first creates a new `ChunkedMemoryStream`, then asynchronously copies the source data into that in-memory stream, and finally performs the rest of the actions on that in-memory stream.  In that latter scenario, synchronous reads are not only fine, but are actually preferrable as they avoid unnecessary overhead.

After thinking about this a bit, I'm now of the opinion that `WithSeekableStreamAsync()` should _always_ copy the image to a `ChunkedMemoryStream`, and then the rest of the internal image decoding logic should use explicitly synchronous code paths, removing any facade of being async at all.  I don't have hard data to back this up, but I suspect that isolating the asynchronous network IO reads to `Stream.CopyToAsync()` would yield better performance than sprinkling those reads throughout the image decoding flow.

NOTE: For the synchronous path, `WithSeekableStream()` could still keep the optimization that skips creating an in-memory stream if the source is seekable.  It would just be in the async path that the code would always copy into an in-memory stream.

**Comment by antonfirsov** [comment]
> When running in a webservice under high-load scenarios, this can quickly lead to thread pool exhaustion as all threads are blocked waiting on network IO.

Do you see this is actually happening? `NetworkStream` is not seekable and I would assume file IO is much faster.

**Comment by kroymann** [comment]
Unfortunately, yes it does.  Over that last few weeks our production service has been encountering sporadic outages that we've tracked down to thread pool exhaustion caused by these reads.  The Azure Blob Storage SDK appears to be returning streams that are seekable:
https://github.com/Azure/azure-storage-net/blob/b0d47bfd4c1c0dcbe60eaf373e2977cd7405aa5f/Lib/Common/Blob/BlobReadStreamBase.cs#L94

**Comment by kroymann** [comment]
I just reviewed the version of the Azure SDK being used by the `ImageSharp.Web.Providers.Azure` library and confirmed that it too returns a Stream that is seekable:
https://github.com/Azure/azure-sdk-for-net/blob/f31090ad65df180fb8b8d07a263ea0fe0c558869/sdk/storage/Azure.Storage.Common/src/Shared/LazyLoadingReadOnlyStream.cs#L355

**Comment by antonfirsov** [comment]
Thanks!

There is no easy solution. Updating decoders to work in a truly async manner is very hard problem as you figured out, I don't think we'll ever go down that rabbit hole. Always prefetching into a `ChunkedMemoryStream` sounds reasonable but that would increase the memory footprint and may put a lot of pressure on `ArrayPool<byte>.Shared`. Probably still the lesser evil. I wonder if we should make it configurable through an option on `Configuration`?

**Comment by kroymann** [comment]
From my vantage point, having an API that advertises itself as being async, but which internally uses synchronous IO seems like a pretty nasty anti-pattern that can't be viewed as anything other than a bug.  All it takes is a short hiccup in network throughput and you can quickly consume every thread in the pool with blocking IO.

Given that the decoders aren't actually async, and as you said it's highly unlikely that anyone will tackle that work, I only see two valid paths to take here:
1. Have the async APIs _always_ copy into an in-memory buffer so that the decoder can then safely execute without incurring blocking IO (and then advertise this behavior as broadly as possible via comments, documentation, etc).
2. Remove the async APIs _entirely_ and instead maybe provide a simple demonstration in the docs of how to use something like a `RecyclableMemoryStreamManager` to asynchronously download an image into memory before calling `Image.Load()`:

```cs
using (var memoryStream = recyclableMemoryStreamManager.GetStream())
{
    await sourceStream.CopyToAsync(memoryStream);
    memoryStream.Position = 0;
    return Image.Load<Rgba32>(configuration, memoryStream);
 }
```

**Comment by kroymann** [comment]
By the way I want to make sure my comments aren't coming off as angry/aggressive/demanding.  I am extremely appreciative of the work that all of the maintainers do on this fantastic product. :bow:  I just want to help ensure that other users of this library don't get bitten by this same issue where they mistakenly believe that `LoadAsync()` will use async IO to read the image.

**Comment by antonfirsov** [comment]
> As it stands right now, we're already in a state where the API is effectively synchronous anyway, so going with the second option isn't really a big loss in functionality - it's merely acknowledging the reality that the image loading code doesn't actually have an async path

This is a very hard breaking change, basically no-go. We should aim to find a fix instead. We should go ahead and prototype a unconditional `ChunkedMemoryStream` preloading. The hardest part of the job is async-enabling [LoadResizeMemoryStress](https://github.com/SixLabors/ImageSharp/blob/main/tests/ImageSharp.Tests.ProfilingSandbox/LoadResizeSaveParallelMemoryStress.cs), and comparing the memory pressure before/after.

@kroymann are you interested in contribution? If yes, can provide more guidance.

@JimBobSquarePants any thoughts? Are you fine with such a change in principle, if it doesn't regress us in other means?

**Comment by kroymann** [comment]
@antonfirsov Happy to contribute (I was already deep down the path of implementing a PR when I filed this issue).  I've been contributing on the ImageSharp.Web project periodically over the years but haven't had a reason until now to dig into ImageSharp itself.

That said, I think I'd still like to make a pitch in favor of the hard breaking change as it feels like a cleaner acknowledgement of the true reality of the code.  The core ImageSharp library simply does not have support for asynchronous image loading, and the existing async APIs with their `Task.FromResult()` wrappers around their synchronous counterparts are just misleading people into thinking that it does.  Rather than bake in a hidden async preload followed by synchronous load, as a consumer I personally would prefer for the library to make it crystal clear to me that it does not support asynchronous loading at all, and then let me add my own async preloading before calling into ImageSharp if that's what I want to do.

I am curious to hear what @JimBobSquarePants thinks on the topic though.

**Comment by antonfirsov** [comment]
>  The core ImageSharp library simply does not have support for asynchronous image loading

This is not true for the path where the stream is pre-fetched to `ChunkedMemoryStream`. It's sub-optimal, but truly async, since there is no blocking IO call.

> and then let me add my own async preloading before calling into ImageSharp if that's what I want to do

The library optimizes this scenario with pooled allocations within `ChunkedMemoryStream`. `MemoryStream` allocates a single large GC array, which creates a lot of GC pressure and hurts throughput.

**Comment by kroymann** [comment]
>The library optimizes this scenario with pooled allocations within `ChunkedMemoryStream`. `MemoryStream` allocates a single large GC array, which creates a lot of GC pressure and hurts throughput.

That's definitely true.  In my product's code, after discovering this problem, we're now working around it by preloading the stream into a [RecyclableMemoryStream](https://github.com/microsoft/Microsoft.IO.RecyclableMemoryStream) which does similar chunking.

**Comment by JimBobSquarePants** [comment]
There's only one option and that's to do the copy. We simply cannot make such a the breaking change. 

To reduce `ArrayPool` pressure I would  replace our own `ChunkedMemoryStream` with `RecyclableMemoryStream`. The only reason I didn't use it originally was because we were targeting netstandard 1.3 

We can then expose the `RecyclableMemoryStreamManager` via `Configuration` as we do in ImageSharp.Web `ImageSharpMiddlewareOptions`

https://github.com/SixLabors/ImageSharp.Web/blob/6912cbc719b6fbab16ff7c8c86e0567e57e0c7c7/src/ImageSharp.Web/Middleware/ImageSharpMiddlewareOptions.cs#L55-L59

I don't think we should make the copy optional and always default to using it.

@kroymann If you could go ahead and give that a go that would be ace! ❤️

**Comment by antonfirsov** [comment]
`RecyclableMemoryStream` would bring an external dependency and create a second parallel set of memory pools. I prefer to avoid this in the core library.

By `ArrayPool` pressure I particularly mean the problem of exhausting a single pool bucket, since `ChunkedMemoryStream` rents uniform buffers. We can try to address it it by incrementally growing `ChunkedMemoryStream`'s `chunkLength` as the stream grows, which is `RecyclableMemoryStream`'s primary trick.

**Comment by JimBobSquarePants** [comment]
Ah right, yep ok, I misunderstood. Keep our implementation for now then and we can look at using the same trick.

**Comment by kroymann** [comment]
Sounds good.  @antonfirsov Any suggestions on how you'd want to tackle doing this:
>The hardest part of the job is async-enabling [LoadResizeMemoryStress](https://github.com/SixLabors/ImageSharp/blob/main/tests/ImageSharp.Tests.ProfilingSandbox/LoadResizeSaveParallelMemoryStress.cs), and comparing the memory pressure with/without preloading to make sure we don't regress significantly.

**Comment by antonfirsov** [comment]
We need 2 new [command line options](https://github.com/SixLabors/ImageSharp/blob/main/tests/ImageSharp.Tests.ProfilingSandbox/LoadResizeSaveParallelMemoryStress.cs): 
- One to make the decoding / encoding async
- One to enable/disable preloading into `ChunkedMemoryStream` in `Image.LoadAsync`. This could be wired into an internal test-only option on `Configuration`.

We then need an async variant of [the actual benchmark code](https://github.com/SixLabors/ImageSharp/blob/f24b02f31c2147a76a8d9894674d889ab194dbbf/tests/ImageSharp.Benchmarks/LoadResizeSave/LoadResizeSaveStressRunner.cs#L179-L198), and [it's call site](https://github.com/SixLabors/ImageSharp/blob/f24b02f31c2147a76a8d9894674d889ab194dbbf/tests/ImageSharp.Tests.ProfilingSandbox/LoadResizeSaveParallelMemoryStress.cs#L313-L331). If the async command line option is set, we should run `ImageSharpBenchmarkParallelAsync().GetAwaiter().GetResult()` instead of the sync variant.

Running & evaluating the benchmarks is [even more tricky](https://github.com/saucecontrol/core-imaging-playground/tree/beeees/MemoryStress) and needs a beefy multi-core test machine, so I'm happy to take that part. I'm just overly paranoid taking more coding work than I can realistically finish in reasonable time.

Ideally it would be great to make this into our 1.0.1 bugfix release, unless @JimBobSquarePants wants to release it ASAP.

**Comment by kroymann** [comment]
I can start on those test improvements.  In the meantime, I've got a [branch](https://github.com/kroymann/ImageSharp/tree/fix-async-image-load) where I've implemented the fix if you'd like to take a look (I can also just start a PR if that's a better way to review things).  In addition to the [simple change](https://github.com/kroymann/ImageSharp/commit/4802b4f0f4c26fc6bbc30266d02734b5269c1600) that ensures we always prebuffer the image stream, I also added two additional commits that I'd like your opinions on.  The [first](https://github.com/kroymann/ImageSharp/commit/3a11b2f3820bc862dbdf55beeae5432e9a5b7ccd) plumbs an optional `CancellationToken` through the `IImageDecoder` and `IImageInfoDetector` methods so that it can be passed into the internal decoder logic.  The [second](https://github.com/kroymann/ImageSharp/commit/f874218d0b5a2cf064aa0a414f333381e9a22127) changes the internal `Image.WithSeekableStreamAsync()` method so that it executes synchronous actions rather than async actions after first prebuffering the image.  This in turn makes it possible to remove some unnecessary async wrappers that weren't actually doing anything asynchronously.

**Comment by JimBobSquarePants** [comment]
@kroymann I'm not sure your [simple change](https://github.com/kroymann/ImageSharp/commit/4802b4f0f4c26fc6bbc30266d02734b5269c1600) is correct there. Rather than deleting the condition you should be copying the stream within the condition. We're seeking to the start of the stream because the configuration property says we should.

I'm kinda ok with binary breaking changes as long as we version accordingly though @antonfirsov might judge me harshly for that 😛

Crack open a PR anyway and we can review properly.

**Comment by kroymann** [comment]
>I'm not sure your [simple change](https://github.com/kroymann/ImageSharp/commit/4802b4f0f4c26fc6bbc30266d02734b5269c1600) is correct there. Rather than deleting the condition you should be copying the stream within the condition. We're seeking to the start of the stream because the configuration property says we should.

I'm not following you here.  That if-statement was the crux of the problem where it was mistakenly assuming that just because the source stream was seekable, that it didn't need to copy the data into an in-memory buffer (and then the decoding process went on to read synchronously from that stream).  The fix is to get rid of that optimization and _always_ copy into the in-memory buffer.

I'll open a PR so we can review and discuss there.

**Comment by kroymann** [comment]
One more thought before I switch the conversation to the PR.  While I think the fix that we're making to the `Image.LoadAsync()` method is OK, if someone were to try and use one of the `IImageDecoder.DecodeAsync()` methods directly rather than going through `Image.LoadAsync()`, then they're still going to hit this issue.  What would you two say to a making a hard breaking change that removes the async methods from `IImageDecoder` and `IImageInfoDetector`?

We'd keep `Image.LoadAsync()` since that's what the majority of people are likely calling, so the impact of the breaking change would hopefully be minimized, but we'd still get rid of `IImageDecoder.DecodeAsync()`, etc?

**Comment by antonfirsov** [comment]
> as long as we version accordingly

I think this is the key. In our ad-hoc versioning system, I believe this means bumping the minor version (-> `1.1.0`), which means *breaking change, but with very impact* - since very few users are implementing their own `IImageDecoder`. We should not merge the PR before releasing the `1.0.1` patch though.

**Comment by JimBobSquarePants** [comment]
I'm beginning to wonder whether we should skip v2.0.1 and release v2.1.0 instead including the changes in the https://github.com/SixLabors/ImageSharp/pull/2006.  

My reasons are as follows.

- I want to include these fixes in the ImageSharp.Web v2.0.0 release.
- It means a users upgrading from ImageSharp < v1.0.4 to the v2 major will not experience a double break in behavior.

**Comment by antonfirsov** [comment]
> I want to include these fixes in the ImageSharp.Web v2.0.0 release.

Why not just wait with the release of ImageSharp.Web until this is fixed? A 2.0.1 release doesn't have to affect it.

> It means a users upgrading from ImageSharp < v1.0.4 to the v2 major will not experience a double break in behavior.

I don't follow this point, an `1.0.*` -> `2.0.0` -> `2.1.0` is already double-breaking path, 2.0.1 won't change it.

---

The only counter-argument IMO is if we want to save time on creating the 2.0.1 release, which can be valid. A factor that might affect the decision is the ETA of #2006. Benchmarking and potential performance mitigations are still ahead.

**Comment by JimBobSquarePants** [comment]
What I mean by my second point is that upgrades to our v2 major will currently be limited. If v2.1.0 including this fix was delayed somehow for a long time then many users will find themselves stuck with a double breaking upgrade in that period. 

Basically I'd like to get anything critical out in a single bump and have user moving from v1.0.4 to the latest and greatest at the earliest opportunity. In addition I'd really like to shift focus to get Drawing and Fonts released.

**Comment by JimBobSquarePants** [comment]
Fixed via #2006 Thanks @kroymann !!
